// react-cellblock/lib/util/validators.js
'use strict';

exports.__esModule = true;
exports.gridFraction = gridFraction;
exports.validBreakpoint = validBreakpoint;
exports.validBreakpoints = validBreakpoints;

var _react = require('react');

var FRACTION_RE = /^\d+\/\d+$/;

function gridFraction(props, propName, componentName) {
  var value = props[propName];
  if (props[propName] && !isFraction(value)) {
    return createError(props, propName, componentName, 'expected a fraction string `a/b` (ie: 2/3)');
  }
}

function validBreakpoint(props, propName) {
  if (typeof window === 'undefined' && !props[propName]) {
    return new Error('Isomorphic grids require an initialBreakpoint');
  }
  return _react.PropTypes.oneOf(props.breakpoints).apply(null, arguments);
}

function validBreakpoints(props, propName, componentName) {
  switch (true) {
    case !Array.isArray(props[propName]) || !isAllIntegers(props[propName]):
      return createError(props, propName, componentName, 'expected an array of integers');
    case !isSorted(props[propName]):
      return createError(props, propName, componentName, 'expected ascending order');
  }
}

function createError(props, propName, componentName, message) {
  var str = 'Invalid prop ' + propName + ' of value ' + props[propName] + ' supplied to ' + componentName;
  return new Error(str + ', ' + message);
}

function isString(v) {
  return typeof v === 'string';
}

function isNumber(v) {
  return typeof v === 'number';
}

function isFraction(v) {
  return isString(v) && FRACTION_RE.test(v);
}

function isInteger(v) {
  return isNumber(v) && v % 1 === 0;
}

function isSorted(arr) {
  for (var i = 0; i < arr.length - 1; i += 1) {
    if (arr[i] > arr[i + 1]) return false;
  }
  return true;
}

function isAllIntegers(arr) {
  for (var i = 0; i < arr.length; i += 1) {
    if (!isInteger(arr[i])) return false;
  }
  return true;
}
